File: OneCode/Visual Studio 2010/CSASPNETCMD/App_Code/BatchRunner.cs
/****************************** Module Header ******************************\
Module Name:  BatchRunner.cs
Project:      CSASPNETCMD
Copyright (c) Microsoft Corporation.

The Batch Runner Class is responsible for executing batch file.

This source is subject to the Microsoft Public License.
See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
All other rights reserved.

THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

#region Using directives
using System;
using System.Security;
#endregion

namespace CSASPNETCMD
{
    public class BatchRunner
    {
        #region fields
        private string domainName;
        private string userName;
        private string password;
        #endregion

        #region properties
        /// <summary>
        /// Set and get identify user's DomainName.
        /// </summary>
        public string DomainName
        {
            get { return domainName; }
            set { domainName = value; }
        }
        /// <summary>
        /// Set and get identify user's UserName.
        /// </summary>
        public string UserName
        {
            get { return userName; }
            set { userName = value; }
        }
        /// <summary>
        /// Set and get identify user's Password.
        /// </summary>
        public string Password
        {
            get { return password; }
            set { password = value; }
        }

        #endregion       

        #region constructor
        /// <summary>
        /// Default constructor
        /// </summary>
        public BatchRunner()
        {
        }

        /// <summary>
        /// Constructor with identify information.
        /// </summary>
        /// <param name="domain">Identify user's DomainName.</param>
        /// <param name="username">Identify user's username.</param>
        /// <param name="password">Identify user's password.</param>
        public BatchRunner(string domain, string username,string password)
        {
            this.domainName = domain;
            this.userName = username;
            this.password = password;
        }

        #endregion

        /// <summary>
        /// Execute batch file.
        /// </summary>
        /// <param name="fileName">The batch file with fullname to execute.</param>
        /// <returns>Result of batch file execution.</returns>
        public string ExecuteBatch(string fileName)
        {
            if (!fileName.ToLower().EndsWith(".bat"))
            {
                throw new ArgumentException("the file's suffix must be .bat");
            }
            string retOutPut = "No Output!";

            try
            {
                System.Diagnostics.ProcessStartInfo psi = GenerateProcessInfo(fileName);

                System.Diagnostics.Process processBatch = System.Diagnostics.Process.Start(psi);
                System.IO.StreamReader myOutput = processBatch.StandardOutput;

                processBatch.WaitForExit();
                if (processBatch.HasExited)
                {
                    retOutPut = myOutput.ReadToEnd();
                }
            }
            catch (System.ComponentModel.Win32Exception winException)
            {
                if (winException.Message.Contains("bad password"))
                {
                    retOutPut = winException.Message;
                }
                else
                {
                    retOutPut = "Win32Exception occured";
                }
                // Log exception information.
            }
            catch (System.Exception exception)
            {
                retOutPut = exception.Message;

                // Log exception information.
            }
            return retOutPut;
        }

        /// <summary>
        /// Generate ProcessInfo with fields(domainName,userName,password).
        /// </summary>
        /// <param name="fileName">The batch file with fullname to execute.</param>
        /// <returns></returns>
        private System.Diagnostics.ProcessStartInfo GenerateProcessInfo(string fileName)
        {
            System.Diagnostics.ProcessStartInfo psi = new System.Diagnostics.ProcessStartInfo(fileName);
            psi.RedirectStandardOutput = true;
            psi.CreateNoWindow = true;
            psi.UseShellExecute = false;
            if (!string.IsNullOrEmpty(domainName))
            {
                psi.Domain = domainName;
            }
            if (!string.IsNullOrEmpty(userName))
            {
                psi.UserName = userName;
            }
            if (!string.IsNullOrEmpty(password))
            {
                SecureString sstring = new SecureString();
                foreach (char c in password)
                {
                    sstring.AppendChar(c);
                }
                psi.Password = sstring;
            }
            return psi;
        } 

    }
}

File: OneCode/Visual Studio 2010/VBASPNETCMD/App_Code/BatchRunner.vb
'***************************** Module Header ******************************\
'Module Name:  BatchRunner.cs
'Project:      CmdOfASPNET
'Copyright (c) Microsoft Corporation.
'
'The Batch Runner Class is responsible for executing batch file.
'
'This source is subject to the Microsoft Public License.
'See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
'All other rights reserved.
'
'THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
'EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
'WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'\**************************************************************************

#Region "Using directives"
Imports System.Security
#End Region

Public Class BatchRunner
#Region "fields"
    Private m_domainName As String
    Private m_userName As String
    Private m_password As String
#End Region

#Region "properties"
    ''' <summary>
    ''' Set and get identify user's DomainName
    ''' </summary>
    Public Property DomainName() As String
        Get
            Return m_domainName
        End Get
        Set(ByVal value As String)
            m_domainName = Value
        End Set
    End Property
    ''' <summary>
    ''' Set and get identify user's UserName
    ''' </summary>
    Public Property UserName() As String
        Get
            Return m_userName
        End Get
        Set(ByVal value As String)
            m_userName = Value
        End Set
    End Property
    ''' <summary>
    ''' Set and get identify user's Password
    ''' </summary>
    Public Property Password() As String
        Get
            Return m_password
        End Get
        Set(ByVal value As String)
            m_password = Value
        End Set
    End Property

#End Region

#Region "constructor"
    ''' <summary>
    ''' Default constructor
    ''' </summary>
    Public Sub New()
    End Sub

    ''' <summary>
    ''' Constructor with identify information
    ''' </summary>
    ''' <param name="domain">Identify user's DomainName</param>
    ''' <param name="username">Identify user's username</param>
    ''' <param name="password">Identify user's password</param>
    Public Sub New(ByVal domain As String, ByVal username As String, ByVal password As String)
        Me.m_domainName = domain
        Me.m_userName = username
        Me.m_password = password
    End Sub

#End Region

    ''' <summary>
    ''' Execute batch file
    ''' </summary>
    ''' <param name="fileName">The batch file with fullname to execute</param>
    ''' <returns>Result of batch file execution</returns>
    Public Function ExecuteBatch(ByVal fileName As String) As String
        If Not fileName.ToLower().EndsWith(".bat") Then
            Throw New ArgumentException("the file's suffix must be .bat")
        End If
        Dim retOutPut As String = "No Output!"

        Try
            Dim psi As System.Diagnostics.ProcessStartInfo = GenerateProcessInfo(fileName)

            Dim processBatch As System.Diagnostics.Process = System.Diagnostics.Process.Start(psi)
            Dim myOutput As System.IO.StreamReader = processBatch.StandardOutput

            processBatch.WaitForExit()
            If processBatch.HasExited Then
                retOutPut = myOutput.ReadToEnd()
            End If
        Catch winException As System.ComponentModel.Win32Exception
            If winException.Message.Contains("bad password") Then
                retOutPut = winException.Message
            Else
                retOutPut = "Win32Exception occured"
                ' Log exception information
            End If
        Catch exception As System.Exception
            ' Log exception information
            retOutPut = exception.Message
        End Try
        Return retOutPut
    End Function

    ''' <summary>
    ''' Generate ProcessInfo with fields(domainName,userName,password)
    ''' </summary>
    ''' <param name="fileName">The batch file with fullname to execute</param>
    ''' <returns></returns>
    Private Function GenerateProcessInfo(ByVal fileName As String) As System.Diagnostics.ProcessStartInfo
        Dim psi As New System.Diagnostics.ProcessStartInfo(fileName)
        psi.RedirectStandardOutput = True
        psi.CreateNoWindow = True
        psi.UseShellExecute = False
        If Not String.IsNullOrEmpty(m_domainName) Then
            psi.Domain = m_domainName
        End If
        If Not String.IsNullOrEmpty(m_userName) Then
            psi.UserName = m_userName
        End If
        If Not String.IsNullOrEmpty(m_password) Then
            Dim sstring As New SecureString()
            For Each c As Char In m_password
                sstring.AppendChar(c)
            Next
            psi.Password = sstring
        End If
        Return psi
    End Function

End Class